### DISCUSSION TITLE
How to ensure regexes in a `RegexSet` are mutually exclusive?

### DISCUSSION THREAD

**Discussion by damooo** [rust-lang/regex#837] [Q&A] [answered]
Thanks for great work.

Is it possible to somehow ensure all regexes in a `RegexSet` are mutually exclusive? i.e. None of haystack must never match more than one regex.
.

Thanks again.

**Comment by BurntSushi** [comment, marked as answer]
It's possible _in theory_, because you can take any two automata and intersect them. If the resulting automata has any reachable match states, then there must be some overlap. Conversely, if there are no reachable match states in the intersection, then the set of strings each regex matches must therefore be disjoint.

But this crate doesn't (and will never) expose the necessary utilities to do such a thing because it's not terribly practical to do in the context of a general purpose regex library. Short of that, I don't know of a way to answer such a query in the general case. If you can assume something about the regexes you're using, then there are likely shortcuts. e.g., `[a-z]{2}` and `[a-z]{3}` are completely disjoint because each regex matches a fixed number of strings of differing lengths. Because each such string must have a different length, they must therefore be disjoint sets.

My guess is that the [XY problem](https://xyproblem.info/) may be at work here. It might be more helpful if you describe the higher level problem you're trying to solve.

**Comment by damooo** [comment]
Thanks @BurntSushi ,

Let's say, we are taking configuration from user for routing url to files. With one capture in url match denoting slug for folder, and other capture denoting file path in matched folder. How to ensure configuration will never lead to overlaps and ambiguities? The folder slug may be sub-domain name or sub path in a domain, etc..

**Reply by BurntSushi** [reply]
I see. Yeah, I think you need automata intersection in order to provide this type of guarantee in the general case. Otherwise, you just have to rely on the user to get it right.

But if all you want to do is remove ambiguity, you can do that by saying that when multiple routes match, the earliest one that is defined is selected.

**Reply by damooo** [reply]
It is needed to reject the config outright, without carrying on non-determinism to run time. Is there any possibility, that one can rely upon for automata intersection? like copying minimal code, if methods are not public, etc ?

**Reply by BurntSushi** [reply]
There is no non-determinism in the regex matching. So if your code is actually non-deterministic, it's coming from your code.

There is nothing in this crate that will help you. You may find the regex-syntax and maybe the regex-automata crates useful. But you'll have to implement (at least) the intersection operation.

**Reply by damooo** [reply]
Thanks for your replies and time @BurntSushi .

Non-determinism, in the sense, about interfacing with user. Or i may say, non-additive config. by adding new rules, previous expectations may be broken. I will check with mentioned crates. I am new to DFAs , but will try.

**Reply by BurntSushi** [reply]
Like I said, you can add a rule that says, "the first route defined that matches will take precedent over other simultaneous matches." Or, at runtime, if more than one match is detected, you can report an error.
